leshan-tl-javacoap-server/src/main/java/org/eclipse/leshan/transport/javacoap/endpoint/JavaCoapServerEndpoint.java

```diff
@@ -0,0 +1,233 @@
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.leshan.core.endpoint.Protocol;
+import org.eclipse.leshan.core.observation.Observation;
+import org.eclipse.leshan.core.observation.SingleObservation;
+import org.eclipse.leshan.core.request.DownlinkRequest;
+import org.eclipse.leshan.core.request.ObserveRequest;
+import org.eclipse.leshan.core.request.exception.SendFailedException;
+import org.eclipse.leshan.core.response.ErrorCallback;
+import org.eclipse.leshan.core.response.LwM2mResponse;
+import org.eclipse.leshan.core.response.ObserveResponse;
+import org.eclipse.leshan.core.response.ResponseCallback;
+import org.eclipse.leshan.server.endpoint.LwM2mServerEndpoint;
+import org.eclipse.leshan.server.endpoint.ServerEndpointToolbox;
+import org.eclipse.leshan.server.observation.LwM2mNotificationReceiver;
+import org.eclipse.leshan.server.profile.ClientProfile;
+import org.eclipse.leshan.server.registration.RegistrationStore;
+import org.eclipse.leshan.server.request.LowerLayerConfig;
+import org.eclipse.leshan.transport.javacoap.observation.ObservationUtil;
+
+import com.mbed.coap.client.CoapClient;
+import com.mbed.coap.client.CoapClientBuilder;
+import com.mbed.coap.client.ObservationConsumer;
+import com.mbed.coap.packet.CoapRequest;
+import com.mbed.coap.packet.CoapResponse;
+import com.mbed.coap.packet.Opaque;
+import com.mbed.coap.server.CoapServer;
+
+public class JavaCoapServerEndpoint implements LwM2mServerEndpoint {
+
+    private final URI endpointUri;
+    private final CoapServer coapServer;
+    private final ServerCoapMessageTranslator translator;
+    private final ServerEndpointToolbox toolbox;
+    private final LwM2mNotificationReceiver notificationReceiver;
+    private final RegistrationStore registrationStore;
+
+    public JavaCoapServerEndpoint(URI endpointUri, CoapServer coapServer, ServerCoapMessageTranslator translator,
+            ServerEndpointToolbox toolbox, LwM2mNotificationReceiver notificationReceiver,
+            RegistrationStore registrationStore) {
+        this.endpointUri = endpointUri;
+        this.coapServer = coapServer;
+        this.translator = translator;
+        this.toolbox = toolbox;
+        this.notificationReceiver = notificationReceiver;
+        this.registrationStore = registrationStore;
+
+    }
+
+    @Override
+    public Protocol getProtocol() {
+        return Protocol.COAP;
+    }
+
+    @Override
+    public URI getURI() {
+        return endpointUri;
+    }
+
+    @Override
+    public <T extends LwM2mResponse> T send(ClientProfile destination, DownlinkRequest<T> request,
+            LowerLayerConfig lowerLayerConfig, long timeoutInMs) throws InterruptedException {
+
+        // Send LWM2M Request
+        CompletableFuture<T> lwm2mResponseFuture = sendLwM2mRequest(destination, request, lowerLayerConfig);
+
+        // Wait synchronously for LWM2M response
+        try {
+            return lwm2mResponseFuture.get(timeoutInMs, TimeUnit.MILLISECONDS);
```

Comment:
How common are blocking API like this? Java-coap was design to be fully asynchronous so blocking calls might cause some issues. Although here it is probably ok.

Reply:
I consider that sync API is mainly useful to write demo, POC or tests.
A production server should rather use the async way.

But I'm curious : 
> Java-coap was design to be fully asynchronous so blocking calls might cause some issues.

which kind of issue are you in mind ?

Reply:
I meant that if while handling a request (for example in Filter) you would block a call, then it might get into deadlock.

Reply:
I get it but I imagine for Future about getting response it is ok, right ? I mean this is one of the future feature.

Reply:
Yes, in this case that's the only way.